config/app/com/microsoft/azure/iotsolutions/uiconfig/services/runtime/KeyVault.java

```diff
@@ -0,0 +1,147 @@
+        String uri = String.format(KEY_VAULT_URI, this.name, secretKey);
+
+        try {
+            return this.keyVaultClient
+                    .getSecret(uri)
+                    .value();
+        } catch (Exception e) {
+            String message = String.format("Failed to get the secret {%s} from the key vault.", secretKey);
+            log.error(message, e);
+            return null;
+        }
+    }
+
+    public boolean hasPath(String key) {
+        return this.keys.stream()
+                .anyMatch(
+                        item -> item
+                                .id()
+                                .toLowerCase()
+                                .contains(key.toLowerCase())
+                );
+    }
+
+    // Get last token of the key for referencing the key vault:
+    // (local setting file ) => key vault denomination
+    // messages.cosmosdb.documentDbConnectionString => documentDbConnectionString
+    private String processSecretKey(String secretKey) {
+        return secretKey.substring(secretKey.lastIndexOf("."));
+    }
+
+    /**
+     * Creates a new KeyVaultCredential based on the access token obtained.
+     *
+     * @return
+     */
+    private ServiceClientCredentials createCredentials() {
+        return new KeyVaultCredentials() {
+
+            //Callback that supplies the token type and access token on request.
+            @Override
+            public String doAuthenticate(String authorization, String resource, String scope) {
+
+                AuthenticationResult authResult;
+                try {
+                    authResult = getAccessToken(authorization, resource);
+                    return authResult.getAccessToken();
+                } catch (Exception e) {
+                    // TODO: Add logging
+                    // e.printStackTrace();
+                }
+                return "";
+            }
+
+        };
+    }
+
+    /**
+     * Private helper method that gets the access token for the authorization and resource depending on which variables are supplied in the environment.
+     *
+     * @param authorization
+     * @param resource
+     * @return
+     * @throws ExecutionException
+     * @throws InterruptedException
+     * @throws MalformedURLException
+     * @throws Exception
+     */
+    private AuthenticationResult getAccessToken(String authorization, String resource) throws InterruptedException, ExecutionException, MalformedURLException {
+
+        AuthenticationResult result = null;
+
+        //Starts a service to fetch access token.
+        ExecutorService service = null;
+        try {
+            service = Executors.newFixedThreadPool(1);
+            AuthenticationContext context = new AuthenticationContext(authorization, false, service);
+
+            Future<AuthenticationResult> future = null;
+
+            //Acquires token based on client ID and client secret.
+            if (this.clientSecret != null && this.clientSecret != null) {
```

Comment:
Should we check for null and empty like c#?

Reply:
done